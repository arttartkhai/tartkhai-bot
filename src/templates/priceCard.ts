//TODO: handle when unlimited supply,
import { FlexBox, FlexBubble, FlexCarousel, FlexMessage } from '@line/bot-sdk'
import { CustomCoinData, PriceCardProp } from 'types/template'

type CreateBox = (data: CustomCoinData) => FlexBox

const isNegative = (value: string) => Number.parseFloat(value) < 0
const getColor = (value: string) => (isNegative(value) ? '#ef4949' : '#008000')
const withArrowAndPercent = (value: string) => (isNegative(value) ? `▼ ${value}%` : `▲ ${value}%`)

const getHeader: CreateBox = (data) => {
  return {
    type: 'box',
    layout: 'horizontal',
    contents: [
      {
        type: 'box',
        layout: 'vertical',
        contents: [
          {
            type: 'text',
            size: 'xl',
            flex: 0,
            contents: [
              {
                type: 'span',
                text: data.name,
              },
              {
                type: 'span',
                text: ` (${data.symbol})`,
              },
            ],
            wrap: true,
            weight: 'bold',
            text: '-',
          },
          {
            type: 'text',
            text: `$${data.price}`,
            color: getColor(data.percentChange_24h),
          },
        ],
        flex: 3,
      },
      {
        type: 'image',
        url: data.logo,
        size: 'xs',
        aspectMode: 'fit',
        flex: 1,
      },
    ],
    flex: 0,
    spacing: 'none',
    margin: 'none',
    paddingBottom: 'none',
  }
}

const getSupplyBox: CreateBox = (data) => {
  const maxSupply = Number.parseInt(data.maxSupply ? data.maxSupply : '')
  const totalSupply = Number.parseInt(data.totalSupply)
  const percentage = Math.floor((totalSupply / maxSupply) * 100).toString()
  return {
    type: 'box',
    layout: 'vertical',
    contents: [
      {
        type: 'text',
        text: 'Supply',
        align: 'center',
      },
      {
        type: 'box',
        layout: 'vertical',
        contents: [
          {
            type: 'box',
            layout: 'vertical',
            contents: [],
            backgroundColor: '#0D8186',
            width: `${percentage}%`,
            height: '15px',
            alignItems: 'center',
            justifyContent: 'center',
          },
          {
            type: 'box',
            layout: 'vertical',
            contents: [
              {
                type: 'text',
                text: `${data.totalSupply} / ${data.maxSupply} (${percentage}%)`,
                color: '#ffffff',
                size: 'xs',
                align: 'center',
              },
            ],
            position: 'absolute',
            height: '15px',
            width: '100%',
          },
        ],
        backgroundColor: '#9FD8E36E',
        height: '15px',
        margin: 'xs',
        position: 'relative',
      },
    ],
  }
}

const getPriceBox: CreateBox = (data) => {
  return {
    type: 'box',
    layout: 'vertical',
    contents: [
      {
        type: 'box',
        layout: 'horizontal',
        contents: [
          {
            type: 'text',
            text: '•',
            weight: 'bold',
            flex: 0,
          },
          {
            type: 'text',
            text: `Rank: ${data.rank}`,
            wrap: true,
            offsetStart: 'xl',
            weight: 'bold',
            size: 'sm',
          },
        ],
        margin: 'sm',
      },
      {
        type: 'box',
        layout: 'horizontal',
        contents: [
          {
            type: 'text',
            text: '•',
            weight: 'bold',
            flex: 0,
          },
          {
            type: 'text',
            text: `MCap: $${data.marketCap}`,
            wrap: true,
            offsetStart: 'xl',
            weight: 'bold',
            size: 'sm',
          },
        ],
        margin: 'sm',
      },
      {
        type: 'box',
        layout: 'horizontal',
        contents: [
          {
            type: 'text',
            text: '•',
            weight: 'bold',
            flex: 0,
          },
          {
            type: 'text',
            text: `Price: $${data.price}`,
            wrap: true,
            offsetStart: 'xl',
            weight: 'bold',
            size: 'sm',
          },
        ],
        margin: 'sm',
      },
      {
        type: 'box',
        layout: 'horizontal',
        contents: [
          {
            type: 'text',
            text: '•',
            weight: 'bold',
            flex: 0,
          },
          {
            type: 'text',
            text: `Volume: ${data.volume}`,
            wrap: true,
            offsetStart: 'xl',
            weight: 'bold',
            size: 'sm',
          },
        ],
        margin: 'sm',
      },
    ],
    paddingTop: 'lg',
  }
}

const getPercentageBox: CreateBox = (data) => {
  return {
    type: 'box',
    layout: 'vertical',
    contents: [
      {
        type: 'box',
        layout: 'vertical',
        contents: [
          {
            type: 'text',
            text: 'Percentage Change',
            weight: 'bold',
          },
        ],
        alignItems: 'center',
      },
      {
        type: 'box',
        layout: 'horizontal',
        contents: [
          {
            type: 'box',
            layout: 'vertical',
            contents: [
              {
                type: 'box',
                layout: 'horizontal',
                contents: [
                  {
                    type: 'text',
                    text: '1h:',
                    flex: 0,
                    size: 'xs',
                  },
                  {
                    type: 'text',
                    text: withArrowAndPercent(data.percentChange_1h),
                    flex: 0,
                    color: getColor(data.percentChange_1h),
                    size: 'xs',
                  },
                ],
                justifyContent: 'space-between',
                alignItems: 'center',
              },
              {
                type: 'box',
                layout: 'horizontal',
                contents: [
                  {
                    type: 'text',
                    text: '24h:',
                    flex: 0,
                    size: 'xs',
                  },
                  {
                    type: 'text',
                    text: withArrowAndPercent(data.percentChange_24h),
                    flex: 0,
                    color: getColor(data.percentChange_24h),
                    size: 'xs',
                  },
                ],
                justifyContent: 'space-between',
              },
              {
                type: 'box',
                layout: 'horizontal',
                contents: [
                  {
                    type: 'text',
                    text: '7d:',
                    flex: 0,
                    size: 'xs',
                  },
                  {
                    type: 'text',
                    text: withArrowAndPercent(data.percentChange_7d),
                    flex: 0,
                    color: getColor(data.percentChange_7d),
                    size: 'xs',
                  },
                ],
                justifyContent: 'space-between',
              },
            ],
          },
          {
            type: 'box',
            layout: 'vertical',
            contents: [
              {
                type: 'box',
                layout: 'horizontal',
                contents: [
                  {
                    type: 'text',
                    text: '30d:',
                    flex: 0,
                    size: 'xs',
                  },
                  {
                    type: 'text',
                    text: withArrowAndPercent(data.percentChange_30d),
                    flex: 0,
                    color: getColor(data.percentChange_30d),
                    size: 'xs',
                  },
                ],
                justifyContent: 'space-between',
              },
              {
                type: 'box',
                layout: 'horizontal',
                contents: [
                  {
                    type: 'text',
                    text: '60d:',
                    flex: 0,
                    size: 'xs',
                  },
                  {
                    type: 'text',
                    text: withArrowAndPercent(data.percentChange_60d),
                    flex: 0,
                    color: getColor(data.percentChange_60d),
                    size: 'xs',
                  },
                ],
                justifyContent: 'space-between',
              },
              {
                type: 'box',
                layout: 'horizontal',
                contents: [
                  {
                    type: 'text',
                    text: '90d:',
                    flex: 0,
                    size: 'xs',
                  },
                  {
                    type: 'text',
                    text: withArrowAndPercent(data.percentChange_90d),
                    flex: 0,
                    color: getColor(data.percentChange_90d),
                    size: 'xs',
                  },
                ],
                justifyContent: 'space-between',
              },
            ],
            flex: 1,
          },
        ],
        spacing: 'xl',
      },
    ],
    paddingTop: 'xl',
  }
}
export const priceCard = (data: CustomCoinData): FlexBubble => {
  return {
    type: 'bubble',
    header: getHeader(data),
    body: {
      type: 'box',
      layout: 'vertical',
      contents: [
        // {
        //   type: 'box',
        //   layout: 'vertical',
        //   contents: [
        //     {
        //       type: 'box',
        //       layout: 'vertical',
        //       contents: [
        //         {
        //           type: 'text',
        //           text: 'Bitcoin (BTC) is a cryptocurrency . Users are able to generate BTC through the process of mining. Bitcoin has a current supply of 18,783,500. The last known price of Bitcoin is 45,616.08914537 USD and is down -0.12 over the last 24 hours ...',
        //           wrap: true,
        //           size: 'xxs',
        //         },
        //       ],
        //     },
        //   ],
        // },
        ...(data.maxSupply ? [getSupplyBox(data)] : []),
        ...[getPriceBox(data)],
        ...[getPercentageBox(data)],
      ],
      paddingTop: 'none',
      paddingStart: 'lg',
      paddingEnd: 'lg',
    },
    footer: {
      type: 'box',
      layout: 'vertical',
      contents: [
        {
          type: 'button',
          style: 'link',
          height: 'sm',
          action: {
            type: 'uri',
            label: 'News',
            uri: 'https://linecorp.com',
          },
        },
        {
          type: 'spacer',
          size: 'sm',
        },
      ],
      flex: 0,
      backgroundColor: '#f8f2dd',
    },
  }
}

export const priceCarousel = (listData: PriceCardProp): FlexMessage => {
  const cardBubble = listData.map((coinData) => priceCard(coinData))
  return {
    type: 'flex',
    altText: 'price flex',
    contents: {
      type: 'carousel',
      contents: cardBubble,
    },
  }
}
