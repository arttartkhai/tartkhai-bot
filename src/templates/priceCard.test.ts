import { priceCard } from './priceCard'

describe('priceCard', () => {
  const mockProp = {
    logo: 'https://s2.coinmarketcap.com/static/img/coins/64x64/1.png',
    marketCap: '846,019,026,100.52',
    // maxSupply: null,
    maxSupply: '21,000,000',
    name: 'Bitcoin',
    percentChange_1h: '0.69',
    percentChange_24h: '4.12',
    percentChange_7d: '7.92',
    percentChange_30d: '37.1',
    percentChange_60d: '31.55',
    percentChange_90d: '-23.47',
    price: '45,049.61',
    rank: '1',
    symbol: 'BTC',
    totalSupply: '18,779,718',
    circulatingSupply: '18,779,718',
    volume: '39,707,372,463.02',
  }

  const testPriceCard = priceCard(mockProp)
  console.log(JSON.stringify(testPriceCard))

  it('should match snapshot for 1 bubble', () => {
    expect(testPriceCard).toMatchSnapshot()
  })
})
