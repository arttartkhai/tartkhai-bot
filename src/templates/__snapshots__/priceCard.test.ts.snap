// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`priceCard should match snapshot 1`] = `
Object {
  "body": Object {
    "contents": Array [
      Object {
        "contents": Array [
          Object {
            "align": "center",
            "text": "Supply",
            "type": "text",
          },
          Object {
            "backgroundColor": "#9FD8E36E",
            "contents": Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#0D8186",
                "contents": Array [],
                "height": "15px",
                "justifyContent": "center",
                "layout": "vertical",
                "type": "box",
                "width": "85%",
              },
              Object {
                "contents": Array [
                  Object {
                    "align": "center",
                    "color": "#ffffff",
                    "size": "xs",
                    "text": "18,779,718 / 21,000,000 (85%)",
                    "type": "text",
                  },
                ],
                "height": "15px",
                "layout": "vertical",
                "position": "absolute",
                "type": "box",
                "width": "100%",
              },
            ],
            "height": "15px",
            "layout": "vertical",
            "margin": "xs",
            "position": "relative",
            "type": "box",
          },
        ],
        "layout": "vertical",
        "type": "box",
      },
      Object {
        "contents": Array [
          Object {
            "contents": Array [
              Object {
                "flex": 0,
                "text": "•",
                "type": "text",
                "weight": "bold",
              },
              Object {
                "offsetStart": "xl",
                "size": "sm",
                "text": "Rank: 1",
                "type": "text",
                "weight": "bold",
                "wrap": true,
              },
            ],
            "layout": "horizontal",
            "margin": "sm",
            "type": "box",
          },
          Object {
            "contents": Array [
              Object {
                "flex": 0,
                "text": "•",
                "type": "text",
                "weight": "bold",
              },
              Object {
                "offsetStart": "xl",
                "size": "sm",
                "text": "MCap: $846,019,026,100.52",
                "type": "text",
                "weight": "bold",
                "wrap": true,
              },
            ],
            "layout": "horizontal",
            "margin": "sm",
            "type": "box",
          },
          Object {
            "contents": Array [
              Object {
                "flex": 0,
                "text": "•",
                "type": "text",
                "weight": "bold",
              },
              Object {
                "offsetStart": "xl",
                "size": "sm",
                "text": "Price: $45,049.61",
                "type": "text",
                "weight": "bold",
                "wrap": true,
              },
            ],
            "layout": "horizontal",
            "margin": "sm",
            "type": "box",
          },
          Object {
            "contents": Array [
              Object {
                "flex": 0,
                "text": "•",
                "type": "text",
                "weight": "bold",
              },
              Object {
                "offsetStart": "xl",
                "size": "sm",
                "text": "Volume: 39,707,372,463.02",
                "type": "text",
                "weight": "bold",
                "wrap": true,
              },
            ],
            "layout": "horizontal",
            "margin": "sm",
            "type": "box",
          },
        ],
        "layout": "vertical",
        "paddingTop": "lg",
        "type": "box",
      },
      Object {
        "contents": Array [
          Object {
            "alignItems": "center",
            "contents": Array [
              Object {
                "text": "Percentage Change",
                "type": "text",
                "weight": "bold",
              },
            ],
            "layout": "vertical",
            "type": "box",
          },
          Object {
            "contents": Array [
              Object {
                "contents": Array [
                  Object {
                    "alignItems": "center",
                    "contents": Array [
                      Object {
                        "flex": 0,
                        "size": "xs",
                        "text": "1h:",
                        "type": "text",
                      },
                      Object {
                        "color": "#008000",
                        "flex": 0,
                        "size": "xs",
                        "text": "▲ 0.69%",
                        "type": "text",
                      },
                    ],
                    "justifyContent": "space-between",
                    "layout": "horizontal",
                    "type": "box",
                  },
                  Object {
                    "contents": Array [
                      Object {
                        "flex": 0,
                        "size": "xs",
                        "text": "24h:",
                        "type": "text",
                      },
                      Object {
                        "color": "#008000",
                        "flex": 0,
                        "size": "xs",
                        "text": "▲ 4.12%",
                        "type": "text",
                      },
                    ],
                    "justifyContent": "space-between",
                    "layout": "horizontal",
                    "type": "box",
                  },
                  Object {
                    "contents": Array [
                      Object {
                        "flex": 0,
                        "size": "xs",
                        "text": "7d:",
                        "type": "text",
                      },
                      Object {
                        "color": "#008000",
                        "flex": 0,
                        "size": "xs",
                        "text": "▲ 7.92%",
                        "type": "text",
                      },
                    ],
                    "justifyContent": "space-between",
                    "layout": "horizontal",
                    "type": "box",
                  },
                ],
                "layout": "vertical",
                "type": "box",
              },
              Object {
                "contents": Array [
                  Object {
                    "contents": Array [
                      Object {
                        "flex": 0,
                        "size": "xs",
                        "text": "30d:",
                        "type": "text",
                      },
                      Object {
                        "color": "#008000",
                        "flex": 0,
                        "size": "xs",
                        "text": "▲ 37.1%",
                        "type": "text",
                      },
                    ],
                    "justifyContent": "space-between",
                    "layout": "horizontal",
                    "type": "box",
                  },
                  Object {
                    "contents": Array [
                      Object {
                        "flex": 0,
                        "size": "xs",
                        "text": "60d:",
                        "type": "text",
                      },
                      Object {
                        "color": "#008000",
                        "flex": 0,
                        "size": "xs",
                        "text": "▲ 31.55%",
                        "type": "text",
                      },
                    ],
                    "justifyContent": "space-between",
                    "layout": "horizontal",
                    "type": "box",
                  },
                  Object {
                    "contents": Array [
                      Object {
                        "flex": 0,
                        "size": "xs",
                        "text": "90d:",
                        "type": "text",
                      },
                      Object {
                        "color": "#ef4949",
                        "flex": 0,
                        "size": "xs",
                        "text": "▼ -23.47%",
                        "type": "text",
                      },
                    ],
                    "justifyContent": "space-between",
                    "layout": "horizontal",
                    "type": "box",
                  },
                ],
                "flex": 1,
                "layout": "vertical",
                "type": "box",
              },
            ],
            "layout": "horizontal",
            "spacing": "xl",
            "type": "box",
          },
        ],
        "layout": "vertical",
        "paddingTop": "xl",
        "type": "box",
      },
    ],
    "layout": "vertical",
    "paddingEnd": "lg",
    "paddingStart": "lg",
    "paddingTop": "none",
    "type": "box",
  },
  "footer": Object {
    "backgroundColor": "#f8f2dd",
    "contents": Array [
      Object {
        "action": Object {
          "label": "News",
          "type": "uri",
          "uri": "https://linecorp.com",
        },
        "height": "sm",
        "style": "link",
        "type": "button",
      },
      Object {
        "size": "sm",
        "type": "spacer",
      },
    ],
    "flex": 0,
    "layout": "vertical",
    "type": "box",
  },
  "header": Object {
    "contents": Array [
      Object {
        "contents": Array [
          Object {
            "contents": Array [
              Object {
                "text": "Bitcoin",
                "type": "span",
              },
              Object {
                "text": " (BTC)",
                "type": "span",
              },
            ],
            "flex": 0,
            "size": "xl",
            "text": "-",
            "type": "text",
            "weight": "bold",
            "wrap": true,
          },
          Object {
            "color": "#008000",
            "text": "$45,049.61",
            "type": "text",
          },
        ],
        "flex": 3,
        "layout": "vertical",
        "type": "box",
      },
      Object {
        "aspectMode": "fit",
        "flex": 1,
        "size": "xs",
        "type": "image",
        "url": "https://s2.coinmarketcap.com/static/img/coins/64x64/1.png",
      },
    ],
    "flex": 0,
    "layout": "horizontal",
    "margin": "none",
    "paddingBottom": "none",
    "spacing": "none",
    "type": "box",
  },
  "type": "bubble",
}
`;
